eXtendable Output Functions as first-class citizen

reneme opened the pull request:
### Motivation

Most PQC schemes use XOFs as a building block in one or the other way. This is a proposal to establish `XOF` as a new public API base class. The only implementation currently provided is `SHAKE-128` and `SHAKE-256`.

### Pull Request Dependencies

* ~#3673~

### Find included

* New top-level module "xof" with `class XOF` as abstract base class
  * `start()` allows priming certain XOFs with a `salt` and `key` (otherwise its a NOOP) 
  * `update()` allows streamed input of an arbitrarily long message
  * `output()` allows streamed access to the XOF's infinitely long output stream
  * typical implementations won't allow `update()`s after the first call to `output()`
* `SHAKE_128_XOF` and `SHAKE_256_XOF` as sub-classes of `XOF`
* `cSHAKE_128_XOF` and `cSHAKE_256_XOF` as sub-classes of `XOF` (for internal use only)
* internal helper functions (roughly) as defined in [NIST SP.800-185](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-185.pdf) Section 2.3
  (those functions can be reused to implement KMAC -- see #3570)

### Things to be considered

#### XOFs based on stream ciphers

Kyber and Dilithium build their "90s-crypto" variants on XOF constructions based on AES-256/CTR. We could look into providing support for such stream-cipher XOFs independent of the CRYSTALS algorithms. (See #3672)

#### Retrofitting PQC algorithms

That's future work and will be done in a follow-up. (See #3672)

reneme commented:
Rebased onto @falko-strenzke's Keccak refactoring as a sanity checks. Findings are in #3673.

randombit commented on src/lib/base/buf_input.h:
```diff
@@ -0,0 +1,73 @@
+      * @param in the input to process as a byte array
+      * @param length of param in in bytes
+      */
+      void update(const uint8_t in[], size_t length) { add_data(in, length); }
```
`Buffered_Computation` is already deprecated, I don't think we should use it for the XOF

This also allows some modernization and minimization of the API. In fact we could literally go to just taking a `std::span<const uint8_t>` and be done with it.

reneme replied:
Mhm, fair enough. Albeit, I found it quite handy to have `update_le()`, for instance.

You're proposing to omit the common base interface altogether, right?

randombit replied:
Exactly so, yes

randombit commented on src/lib/hash/shake/shake.h:
```diff
@@ -9,71 +10,70 @@
+* For variable output lengths consider using the XOF interface
 */
-class SHAKE_256 final : public HashFunction {
+class SHAKE_128 final : public detail::SHAKE_Adapter<SHAKE_128_XOF, SHAKE_128> {
```
All of this seems completely unnecessary. Why not just hold a `std::unique_ptr<XOF>` of the appropriate type?

reneme replied:
The rationale is similar to [the one discussed back in May](https://github.com/randombit/botan/pull/3525#discussion_r1186276249). I think, we never really came to a conclusion whether this is just a code smell or might cause actual trouble.

Regardless: I decided not to touch the `HashFunction` implementation of SHAKE in this PR at all. With #3675 introducing `Keccak_Permutation::squeeze()`, there's no significant difference in level of abstraction between `Keccak_Permutation` and `SHAKE_XOF`. Hence, I think it's just fine to avoid the indirection via `SHAKE_XOF` here.

randombit commented on src/lib/stream/shake_cipher/shake_cipher.h:
```diff
@@ -1,77 +1,128 @@
+
+         const auto block_size = m_keystream_buffer.size();
+
+         auto cipher_some = [&](size_t bytes) {
```
I don't see why this logic has to be in a header

reneme replied:
Yeah, that was still sketchy. Addressed in #3675.

Now that `Keccak_Permutation::squeeze()` can handle multiple calls to pull the sponge output in arbitrarily-sized chunks, I decided not the base `SHAKE_Cipher` on the `SHAKE_XOF` but directly on `Keccak_Permutation`. There's no difference in level of abstraction anyway.

reneme commented:
The whole Keccak-Refactoring is becoming a mess with several inter-dependent pull requests. Sorry for that. Lets put this one on-hold until #3673 and #3675 are integrated.

reneme commented:
Re: XOFs based on stream ciphers

Maybe we should step back and rethink the XOF interface somewhat. Looking at [NIST SP.800-185](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-185.pdf) they define things like KMAC-XOF that is KMAC with a variable output length. I.e. just like the AES/CTR-XOF used in CRYSTALS Kyber/Dilithium, it takes a key (and potentially other customization parameters).

I think it would be great to incorporate such XOFs off the bat, when introducing a new API.

reneme commented:
To my mind an abstract XOF should have two phases:

1. Parameterization
2. Output

The "output phase" is simple and inherently generic. The user just calls `::output()` as often as they want to read the indefinite stream of XOF output bytes.

The "parameterization phase" is algorithm-specific, unfortunately. Here are some relevant examples:

| Algorithm   | Parameters                                |
| ----------- | ----------------------------------------- |
| SHAKE       | an arbitrary-length message `msg`         |
| cSHAKE      | arbitrary-lengths `N` and `S`, and `msg`  |
| AES/CTR-XOF | a short `iv` and `key`                    |
| KMAC-XOF    | arbitrary-length `key`, `msg` and `salt`  |

Reflecting this parameter space in a common `XOF` base-interface is possible but probably won't result in a convenient API. Except, if the only "common" feature of this interface is the `::output()` method which could be shared by all XOFs.

reneme commented:
Given the above-described constraints, it seems reasonable to abandon the ubiquitous `::create("AlgoSpec")` factory methods for XOFs. Instead, the user could instantiate the concrete XOF subclass, parameterize it and then (maybe) use it via the abstract `XOF` base class that just provides the `::output()` functionality.

@randombit Does that sound like a reasonable abstraction to you?

reneme commented:
Benchmarking the SHAKE-128/256 XOF showed comparable results for `update()` and `output()` and correspond to [the measured performance of SHA-3(256)/SHA-3(512)](https://github.com/randombit/botan/pull/3673#issuecomment-1691130234). No obvious bottlenecks.

|            | absorb()    | squeeze()   |
| ---------- | ----------- | ----------- |
| SHAKE-128  | 555MiB/sec  | 530MiB/sec  |
| SHAKE-256  | 448MiB/sec  | 432MiB/sec  |

randombit commented:
>  Instead, the user could instantiate the concrete XOF subclass, parameterize it and then (maybe) use it via the abstract XOF base class that just provides the ::output() functionality.

I very much dislike this. A lot of effort was spent in 2->3 in removing all of the concrete subclasses from the public API; this minimizes API and ABI surface. This would also complicate the FFI interface.

I don't think CTR mode "XOF" is worth optimizing for. It is a weird corner case that should not be exposed to applications. And it seems Kyber is dropping "90s mode" entirely (#3543) so it is probably going to be removed soonish anyway.

Then we're left with 

* SHAKE: arbitrary length msg
* cSHAKE: arbitrary length msg. `N` is not for us - quoting SP 800-185 "N is a function-name bit string, used **by NIST** to define functions based on cSHAKE. When no function other than cSHAKE is desired, N is set to the empty string." and so can be ignored. And `S` is a domain separator and can be instantiated in the usual way eg `cSHAKE(256, my_domain_dep)`
* KMAC-XOF: arbitrary length msg, key, and salt.

```
class XOF {
  virtual void start(std::span<const uint8_t> salt = {}, std::span<uint8_t> key = {}) = 0;
  virtual bool valid_salt_length(size_t salt_len) = 0;
  // returns Key_Length_Specification(0) if key is not supported
  virtual Key_Length_Specification key_spec() const = 0;
  virtual void update(std::span<const uint8_t> input);
  //...
};
```

CTR, which doesn't support inputs, is handled by the funny case that `XOF::update` throws `Not_Implemented`.

reneme commented:
@randombit Thanks for your opinion. Your points make sense and I agree in general. I'll be AFK until the beginning of September and will come back to this after.

randombit commented:
Needs a rebase now that the SHA-3 refactor is merged; I'll review after.

reneme commented:
> Needs a rebase now that the SHA-3 refactor is merged; I'll review after.

Rebased. I hope, I get time today to work on the extended API [discussed here](https://github.com/randombit/botan/pull/3671#issuecomment-1691595134).

reneme commented on src/tests/data/xof/shake.vec:
```diff
@@ -0,0 +1,539 @@
+# These vectors are selected from the NIST CAVS file for SHAKE
+#  Generated on Thu Jan 28 14:46:49 2016
+
+[SHAKE-128]
```
Note to self: the Keccak permutation provides a `bmi2` implementation. Probably, we want to insert

```
#test cpuid bmi2
```

in that test vector file to exercise the alternative provider.

randombit commented on src/lib/xof/shake_xof/shake_xof.cpp:
```diff
@@ -0,0 +1,43 @@
+   m_output_generated = false;
+}
+
+void SHAKE_XOF::add_data(const uint8_t key[], size_t length) {
```
Calling this param `key` seems odd since there is no requirement that XOF input be secret, and is inconsistent with the other declarations which call this `input` which seems a better name to capture the non-specific nature of the bytes here.

reneme replied:
For sure. Not sure where this came from. Perhaps a copy & paste quirk. Good catch.

reneme commented:
For tomorrow:

* [ ] Revert the `Buffered_Input` refactoring (see https://github.com/randombit/botan/pull/3671#discussion_r1297073573)
* [ ] Rebase #3672 and try the extended API (`XOF::start()`)
* [ ] Look into cSHAKE

reneme commented on src/lib/xof/xof.h:
```diff
@@ -0,0 +1,196 @@
+
+      /**
+       * Add @p input data to the XOF's internal state
+       *
+       * @param input  the data that shall be
+       */
+      void update(std::span<const uint8_t> input) { add_data(input); }
+
+      /**
+       * @return the next @p bytes output bytes as the specified container type @p T.
+       */
+      template <concepts::resizable_byte_buffer T = secure_vector<uint8_t>>
+      T output(size_t bytes) {
+         T out(bytes);
+         generate_bytes(out);
+         return out;
+      }
+
+      /**
+       * Convenience overload to generate a std::vector<uint8_t>. Same as calling
+       * `XOF::output<std::vector<uint8_t>>()`.
+       *
+       * @return the next @p bytes output bytes as a byte vector.
+       */
+      std::vector<uint8_t> output_stdvec(size_t bytes) { return output<std::vector<uint8_t>>(bytes); }
+
+      /**
+       * Fill @p output with the next output bytes. The number of bytes
+       * depends on the size of @p output.
+       */
+      void output(std::span<uint8_t> output) { generate_bytes(output); }
+
+      /**
+       * @return the next single output byte
+       */
+      uint8_t output_next_byte() {
+         uint8_t out;
+         generate_bytes({&out, 1});
+         return out;
+      }
```
`XOF` does not inherit from a generic `Buffered_Input` class anymore. Instead it directly defines `update(std::span<>)` and no additional convenience methods. For consistency, we should probably remove the convenience methods of `output()` as well, no?

My two cents: I did enjoy having some convenience overloads (particularly taking `std::string_view` and Big/Little-Endian integers). Re-implementing them here as well as later in `HashFunction` and `MessageAuthenticationCode` feels like violating the DRY principle, though. Hence, I wouldn't do that.

Nevertheless, I'd suggest keeping two flavors of `output()`:

* `void output(std::span<>)`
   to allow the user to handle memory allocation
* `T output(size_t bytes)` with `T` being a user-defined container that defaults to `secure_vector<>`
   to let the user conveniently pull data wrapped in their favorite vector-ish.

randombit reviewed (commented):
First pass review

reneme commented on src/lib/xof/shake_xof/shake_xof.cpp:
```diff
@@ -0,0 +1,118 @@
+      // We deliberately do not check the salt length here. There's no technical
+      // reason to reject a salt buffer passed in by an application that exceeds
+      // the (arbitrarily set) valid_salt_length() restriction.
+      keccak_absorb_padded_strings_encoding(*this, block_size(), m_function_name, salt);
```
Yikes, indeed! The condition above was previously meant to exclude non-cSHAKE instances of this base class. But I must have broken it in some restructuring. Anyway, that's certainly not optimal.

I'll look into giving the cSHAKE variants their own base class and leave the salt logic out of the vanilla SHAKE implementation.

reneme commented on src/lib/xof/shake_xof/shake_xof.cpp:
```diff
@@ -0,0 +1,118 @@
+
+cSHAKE_128_XOF::cSHAKE_128_XOF(std::string_view function_name) :
+      cSHAKE_128_XOF(std::vector<uint8_t>{cast_char_ptr_to_uint8(function_name.data()),
+                                          cast_char_ptr_to_uint8(function_name.data()) + function_name.size()}) {}
```
I'd love that. Unfortunately, the base-class constructor additionally takes the `capacity`. This value is hard-coded in the concrete overloads in cSHAKE-128/256 classes. Therefore we can't just conveniently use the delegates in the derived classes. 😞

reneme replied:
Nevertheless, moving the data type conversions into the base class helps a little bit. Then the derived class constructors simply delegate to their base-class counter part.

reneme commented on src/lib/xof/xof.h:
```diff
@@ -0,0 +1,196 @@
+      virtual Key_Length_Specification key_spec() const = 0;
+
+      /**
+       * @return the intrinsic processing block size of this XOF
```
I changed this to be a pure-virtual, forcing downstream implementations to make up their minds.

reneme commented on src/tests/test_xof.cpp:
```diff
@@ -0,0 +1,136 @@
+            // after calling `XOF::clear()`.
+            const auto new_accepts_input = xof->accepts_input();
+
+            result.confirm("advertised block size is > 0", xof->block_size() > 0);
```
[Yes](https://github.com/randombit/botan/pull/3671#discussion_r1321161215).

reneme commented on src/lib/xof/xof.h:
```diff
@@ -0,0 +1,196 @@
+      /**
+       * Some XOFs can be parameterized with a @p salt and/or @p key. If required,
+       * this must be called before calling XOF::update(). XOFs that don't support
+       * either of these parameters will simply ignore it silently.
```
Changed as described [here](https://github.com/randombit/botan/pull/3671#discussion_r1321161215).

reneme commented on src/lib/permutations/keccak_perm/keccak_helpers.cpp:
```diff
@@ -0,0 +1,55 @@
+namespace {
+
+size_t int_encoding_size(uint64_t x) {
+   return ceil_tobytes(std::max(uint8_t(1), ceil_log2(x + 1)));
```
True, that'll result in a wrong result. The int-encodings are mostly used for encoding bitstring-lengths (for salts, function names and friends). I.e. in practice that shouldn't ever happen. I added an assert to be on the safe side, though. Good catch!

reneme commented:
Thanks for the review! I rebased to latest master (after merging the `std::span` refactorings) and addressed your review comments.

Most notably, I split the `SHAKE_XOF` and `cSHAKE_XOF` into two separate base classes (and modules). That results in a handful duplicated lines (`::add_data()`, `::generate_bytes()`) but splits off the salt handling in `::start()` much more logically.

randombit reviewed (commented):
Looks good now. Should we have some tests for cSHAKE as well?

falko-strenzke commented:
There are four [test vectors from NIST](https://csrc.nist.gov/CSRC/media/Projects/Cryptographic-Standards-and-Guidelines/documents/examples/cSHAKE_samples.pdf), all of which have an empty value for N.

I just implemented cSHAKE for another libgcrpyt and created two test vectors with non empty N using  https://asecuritysite.com/golang/cs. I checked their results for non-empty N and that is fine.

```
  { /* Created with https://asecuritysite.com/golang/cs */
    GCRY_MD_CSHAKE128,
    "00010203",
    "ABC",
    "Email Signature",
    32,
    "5CF74DC523ADC0B97EC3614E703835277E9F818879AA1EAE5B2B4E4472EB6A68" },
  { /* Created with https://asecuritysite.com/golang/cs */
   CSHAKE256,
    "00010203", // data hex
    "ABC", // N as C string
    "Email Signature", // S as C string
    32, // output size in bytes
    "0C34C14C4A56E5FC01BE8C04C759DA61437E86B88DF3E21A934436D427A85E9D"  // expected output hex
}, 
```

randombit commented:
Thanks for checking this @falko-strenzke. @reneme can we get these included here so we have some baseline check on cSHAKE? I guess it is a little complicated by cSHAKE not being exposed to applications so you can't immediately use the normal XOF test.

reneme commented:
@falko-strenzke Nice! Thanks for the test vectors. I did look for something "official" for cSHAKE but wasn't successful.

@randombit I hacked the XOF allocation in `test_xof.cpp` so that it explicitly deals with the internal cSHAKE XOFs when needed. Similarly, I also added vectors for the AES/CTR-XOF in #3672 a while ago.

randombit commented:
@reneme looks good thanks. I couldn't find the tests themselves though - are you missing a `git add` of `cshake.vec`?

reneme commented:
I was fighting some linker visibility issues. Had to move the `cSHAKE_XOF::provider()` and `::block_size()` implementations into the *.cpp file to avoid exposing the `Keccak_Permutation` in the shared library.

Now everything should turn green. ☺️

reneme commented:
# 🥳
